coreneuron/io/phase1.cpp

```diff
@@ -0,0 +1,170 @@
+            nrn_assert(srcgid < maxgid);
+            srcgid += zz;
+        }
+    }
+}
+
+void Phase1::add_extracon(NrnThread& nt, int imult) {
+    if (nrn_setup_extracon <= 0) {
+        return;
+    }
+
+    // very simplistic
+    // Use this threads positive source gids - zz in nt.netcon order as the
+    // source gids for extracon.
+    // The edge cases are:
+    // The 0th duplicate uses uses source gids for the last duplicate.
+    // If there are fewer positive source gids than extracon, then keep
+    // rotating through the nt.netcon .
+    // If there are no positive source gids, use a source gid of -1.
+    // Would not be difficult to modify so that random positive source was
+    // used, and/or random connect to another duplicate.
+    // Note that we increment the nt.n_netcon at the end of this function.
+    int sidoffset = 0;  // how much to increment the corresponding positive gid
+    // like ring connectivity
+    if (imult > 0) {
+        sidoffset = -maxgid;
+    } else if (nrn_setup_multiple > 1) {
+        sidoffset = (nrn_setup_multiple - 1) * maxgid;
+    }
+    // set up the extracon srcgid_
+    int* nc_srcgid = netcon_srcgid[nt.id];
+    int j = 0;  // rotate through the n_netcon netcon_srcgid
+    for (int i = 0; i < nrn_setup_extracon; ++i) {
+        int sid = -1;
+        for (int k = 0; k < nt.n_netcon; ++k) {
+            // potentially rotate j through the entire n_netcon but no further
+            sid = nc_srcgid[j];
+            j = (j + 1) % nt.n_netcon;
+            if (sid >= 0) {
+                break;
+            }
+        }
+        if (sid < 0) {  // only connect to real cells.
+            sid = -1;
+        } else {
+            sid += sidoffset;
+        }
+        nc_srcgid[nt.n_netcon + i] = sid;
+    }
+    // finally increment the n_netcon
+    nt.n_netcon += nrn_setup_extracon;
+}
+
+void Phase1::populate(NrnThread& nt, int imult) {
+    nt.n_presyn = this->output_gids.size();
+    nt.n_netcon = this->netcon_srcgids.size();
+
+    shift_gids(imult);
+
+    netcon_srcgid[nt.id] = new int[nt.n_netcon + nrn_setup_extracon];
+    std::copy(this->netcon_srcgids.begin(), this->netcon_srcgids.end(),
+              netcon_srcgid[nt.id]);
+
+    nt.netcons = new NetCon[nt.n_netcon + nrn_setup_extracon];
+    nt.presyns_helper = (PreSynHelper*)ecalloc_align(nt.n_presyn, sizeof(PreSynHelper));
+
+    nt.presyns = new PreSyn[nt.n_presyn];
+    PreSyn* ps = nt.presyns;
+    for (auto& gid: this->output_gids) {
+        if (gid == -1) {
+            ++ps;
+            continue;
+        }
+
+        // Note that the negative (type, index)
+        // coded information goes into the neg_gid2out[tid] hash table.
+        // See netpar.cpp for the netpar_tid_... function implementations.
+        // Both that table and the process wide gid2out table can be deleted
+        // before the end of setup
+
+        /// Put gid into the gid2out hash table with correspondent output PreSyn
```

Comment:
what happened to `MUTLOCK` here?

This reminded me : https://github.com/BlueBrain/CoreNeuron/issues/292

Reply:
To not propagate the global `_mut` to this file, `MUTLOCK`/`MUTUNLOCK` is on the whole function now.

Reply:
Yes, I know this is a trade off. Otherwise expose the mutex mechanisms is a bit hard.

Reply:
Do you want that I did this one or no?

Reply:
Lets keep that as you have done now..would be good if you create a separate issue for this.

Reply:
I try to reduce it. Tell me what you think about it in a new message